astropy/cosmology/funcs.py

```diff
@@ -113,6 +151,16 @@ def z_at_value(func, fval, zmin=1e-8, zmax=1000, ztol=1e-8, maxfun=500):
     other commonly inverted quantities) are monotonic in flat and open
     universes, but not in closed universes.
     """
+    is_array = (
+           (isinstance(fval, Quantity) and not fval.isscalar) or
+           (not isinstance(fval, Quantity) and isinstance(fval, np.ndarray))
+           )
+    if is_array:
+        return _z_at_array(func, fval,
+                           zmin=zmin, zmax=zmax,
+                           nbins=nbins, logspace=logspace,
+                           interpolation=interpolation)
+
```

Comment:
Again, line length limits are not that strict anymore.
Note that the `is_array` condition from above would include length-1 arrays!

This raises some more general issues on efficiency, but if they cannot be addressed (all) now, as already noted in the discussion, this should not keep this PR from being accepted as a first implementation:

1. Default `nbins` could be adapted to the provided `zmin-zmax` range. Some general tests on the required grid resolution would probably be helpful, considering the original example was only using 50. I think I had made some; don't recall the results, but 10000 is certainly very generous.
2. `zmin`, `zmax` themselves might be adapted to the range required for the input `fval`. For monotonic functions this would be straightforward, using `fvals.min()`, `fvals.max()` as already demonstrated in the old example. For the non-monotonic ones this would of course become a bit trickier, like always including the function maximum and then finding the minimum values on both sides. For arbitrary functions and cosmology realisations not that easy...
3. The typical direct minimisation takes O(20) function calls with the new solvers; for small arrays direct calls to the solver for each `fval` will therefore be much more efficient. As a guide, one `z_at_value(func, fval[i])` call is something like 40-100 time faster than the `func(fval)` call on an array[10000].

Reply:
Quick comment re: 3 (I agree with other points): evaluating on a grid of 1e4 values is not necessarily (1e4/20)x slower than direct minimisation with 20 steps, since evaluating on a grid is trivially vectorizable, whereas optimization is serial. If the underlying operations are vectorized well, the grid evaluation + interpolation might even be faster than direct minimisation for a single point! But worth testing.